Add blog post announcing actual dockershim removal (#30972)
kubernetes/website: content/en/blog/_posts/2021-12-16-kubernetes-is-moving-on-from-dockershim.md

---
layout: blog
title: "Kubernetes is moving on from Dockershim: commitments and next steps"
date: 2021-12-16
slug: kubernetes-is-moving-on-from-dockershim
---

**Authors:** Sergey Kanzhelev (Google), Jim Angel (Google), Davanum Srinivas (VMware), Shannon Kularathna (Google), Chris Short (AWS), Dawn Chen (Google)

Kubernetes is removing dockershim in upcoming version 1.24. Please welcome new
container runtimes with more exciting features, a smaller kubelet, and faster

> engineering velocity for new features. If you [use Docker as a Container Runtime](https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/find-out-runtime-you-use/)

Reviewer: ```suggestion
engineering velocity for new features. If you [use Docker as a Container Runtime](/docs/tasks/administer-cluster/migrating-from-dockershim/find-out-runtime-you-use/)
```

for Kubernetes, get ready to migrate from dockershim! To check if you're

> affected, refer to [Check whether dockershim deprecation affects you](https://kubernetes.io/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/).

Reviewer: ```suggestion
affected, refer to [Check whether dockershim deprecation affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/).
```

> ## Why we’re moving away from dockershim

Reviewer 2: This is covered in line 20 but maybe could be higher up so it's more prominent. Maybe add to the intro paragraph for the post like:

Kubernetes is removing dockershim in upcoming version 1.24.  Dockershim was a container runtime interface (CRI) that let system components (like the kubelet) talk to the Docker container runtime. With the dockershim removal, we're welcoming container runtimes with more exciting features, a smaller kubelet, and faster
engineering velocity for new features."?

> ## Why we’re moving away from dockershim

Reviewer 3: Agree on using more words, I think @shannonxtreme's wording is just the right amount of detail.

Docker was the first container runtime used by Kubernetes. This is one of the
reasons why Docker is so familiar to many Kubernetes users and enthusiasts.
Kubernetes uses dockershim as a container runtime interface (CRI) to let system
components (like the kubelet) talk to Docker. It was built into Kubernetes code.
Removing dockershim inevitably affects the entire CNCF family of projects beyond
Kubernetes. Dependencies on Docker and dockershim have crept into various tools
and projects in the Kubernetes ecosystem.

> There are a few container runtimes today that are often lighter, faster and more

Reviewer: Nit. Oxford comma.

There are a few container runtimes today that are often lighter, faster, and more secure than when Docker was the only choice

secure than when Docker was the only choice. By removing the dockershim CRI,
we're embracing the first value of CNCF: "[Fast is better than slow](https://github.com/cncf/foundation/blob/master/charter.md#:~:text=Fast%20is%20better%20than%20slow)".
Migrating to another runtime will ensure that applications are more portable,
with fewer dependencies on the infrastructure. Your apps will be more secure,
with new features unlocked. A different runtime will also let us develop

> features like cgroupv2 support faster.

Author: just remove the last sentence?

> features like cgroupv2 support faster.

Reviewer 4: I'd shorten the text here considerably.

> features like cgroupv2 support faster.

Reviewer 5: I'd delete lines 28-34 and replace with (sorry, @SergeyKanzhelev  the suggest feature isn't working correctly): 

"There are a number of technical, historical, and community-oriented reasons why we think this is the best move for the Kubernetes project that we'll explore in detail in a future post. Stay tuned for future communications on the topic!"

## Deprecation timeline

> We announced dockershim deprecation in December 2020.  Full removal is targeted

Reviewer: Possible rephrase & linking to actual announcement:
```suggestion
We announced [dockershim deprecation](/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation) in December 2020 as part of the 1.20 Kubernetes release. Full removal is targeted 
```

> to Kubernetes 1.24, which is planned for April 2022. This deprecation timeline

Reviewer 5: ```suggestion
in Kubernetes 1.24, in April 2022. This timeline
```

> aligns with our [deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior),

Reviewer: ```suggestion
aligns with our [deprecation policy](/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior),
```

which states that deprecated behaviors must function for no less than 1 year

> after their announced deprecation. We'll support version 1.23, which has

Reviewer 5: ```suggestion
after their announced deprecation. The Kubernetes project will still maintain and support Kubernetes 1.23, release with
```

Reviewer 2: Minor suggestion: 

The Kubernetes project will maintain and support the Kubernetes 1.23 release, with dockershim, until 1.23 reaches its [end of life](/releases/#release-v1-23) on February 28, 2023

dockershim, for another year in the Kubernetes project (and likely longer with

> managed Kubernetes providers). You'll have time to migrate. If you have more

Reviewer 5: ```suggestion
```suggestion
dockershim, until its [end of life](/releases/#release-v1-23), February 28, 2023. 
Managed Kubernetes providers will likely provide longer support periods.
If you have more
```

> questions about the dockershim removal, refer to the [Dockershim Deprecation FAQ](https://k8s.io/dockershim).

Reviewer: ```suggestion
questions about the dockershim removal, refer to the [Dockershim Deprecation FAQ](/blog/2020/12/02/dockershim-faq/).
```

We asked you whether you feel prepared for the migration from dockershim in this

> survey: [Are you ready for Dockershim removal](https://kubernetes.io/blog/2021/11/12/are-you-ready-for-dockershim-removal/).

Reviewer: ```suggestion
survey: [Are you ready for Dockershim removal](/blog/2021/11/12/are-you-ready-for-dockershim-removal/).
```

We had about 600 responses. To everybody who took time filling out the survey,
thank you.

The results show that we still have a lot of ground to cover to help you to

> migrate smoothly. Several newer runtimes exist, and have been promoted

Reviewer 5: ```suggestion
migrate smoothly. Other container runtimes exist, and have been promoted
```

let's try and avoid making value statements on Docker itself (i.e. that it's "old").  Remember that anything we say in a blog post like this becomes the opinion of the project - we don't want our opinion to reflect poorly on Docker.

extensively. However, a significant number of users still rely on dockershim,
and sometimes have dependencies that need to be re-worked. Some of these
dependencies are outside of your control. Based on the feedback received from

> you, here are some of the steps we are taking to help.

Reviewer 5: ```suggestion
extensively. However, many users still rely on dockershim,
and sometimes have dependencies that need to be re-worked. Some of these
dependencies are outside of your control. Based on the feedback received from
you, here are some of the steps we are taking to help.
```

## Our next steps

> Based on the feedback you provided, we'll be doing the following:

Reviewer 5: ```suggestion
Based on the feedback you provided:
```

> - CNCF will sponsor work on updating documentation on how to migrate from

Reviewer 2: [nit]

Do we want to link to the Github project tracking docs work, so that it's clear this work is already underway? Something like "To check the ongoing work, refer to [Docs - Dockershim Removal]\(https://github.com/orgs/kubernetes/projects/67)."?

Reviewer 5: No, don't do that. :)

> - CNCF will sponsor work on updating documentation on how to migrate from
>   dockershim. We'll use survey feedback and our own experiences migrating from
>   dockershim to, try and cover as many use cases and known issues as possible.
> - We have begun, and will continue to, reach out to various projects and
>   vendors to make sure there is a path for customers migration.
> - The release team for Kubernetes 1.24 has set a high bar for documentation
>   readiness.

Reviewer 5: ```suggestion
- CNCF and the 1.24 release team are committed to delivering documentation in time for the 1.24 release. This includes a series of blog posts like this one, giving you timely advice on what upgrading to 1.24 entails, updating existing code samples, tutorials, and documentation, and producing a migration guide for concerned cluster operators. 
- We are reaching out to the rest of the CNCF community to help prepare and ready them for this change
- We are working with Mirantis and Docker to ensure that there is some path to comparability in the future.
```

cc @mrbobbytables and @amye - can we say that last bullet point?

Author: this may be a false promise that this will be path to compatibility. Some compatibility issues will be resolved, but not all.

Reviewer 2: - CNCF and the 1.24 release team are committed to delivering documentation in time for the 1.24 release. This includes more informative blog posts like this one, updating existing code samples, tutorials, and tasks, and producing a migration guide for cluster operators. 
- We are reaching out to the rest of the CNCF community to help prepare them for this change.

Reviewer 5: A suggest with @shannonxtreme's wording, which is 👍 

```suggestion
- CNCF and the 1.24 release team are committed to delivering documentation in time for the 1.24 release. This includes more informative blog posts like this one, updating existing code samples, tutorials, and tasks, and producing a migration guide for cluster operators.
- We are reaching out to the rest of the CNCF community to help prepare them for this change.
```

> If you're part of a project with dependencies on dockershim, or if you're
> interested in helping with the migration effort, please join us! There's always
> room for more contributors, whether to our transition tools or to our
> documentation. To get started, say hello in

Reviewer 5: ```suggestion
If you want to see the changes for yourself, pull down an alpha release here: LINK TBD. 

If you want to get involved, the documentation team would love your help! Drop into [#sig-docs-dockershim](https://kubernetes.slack.com/archives/C02PZMA3R6E) where you can help migrate existing code samples to containerd, update diagrams, and more.
```

Author: not sure what changes? Do you mean 1.24 beta release?

Reviewer 4: Maybe let's not signpost people to that Slack channel at all? If we wanted to, I'd put an article up on https:/k8s.dev/blog/ instead.

> [#sig-docs-dockershim](https://kubernetes.slack.com/archives/C02PZMA3R6E)

Reviewer 5: I would remove this. I'm sorry, but I don't want that channel turning into a general free-for-all for dockershim questions and answers. that channel needs to remain focused on the _actual work_ that needs doing around comms and docs.

@mrbobbytables - can we create a `#dockershim-faq` channel on Slack to funnel people into instead?

Author: good idea

Author: for now changed to `#sig-node`. Maybe even keep it instead of creating a new channel

channel on [CNCF Slack](http://slack.cncf.io/)!

> ## Considerations we've made

Reviewer 5: ```suggestion
## Final thoughts
```

---
layout: blog
title: "Kubernetes is moving on from Dockershim: commitments and next steps"

> date: 2021-12-16

Reviewer 4: Please update this (and the path) to publish on either:
2021-12-20 (preferred)

2021-12-22
2021-12-19

Reviewer 4: The motivation is to separate it from the Kubernetes post-release announcements for v1.23

Reviewer 5: @sftim - another option is publish this January 3rd, 2021, and kick off the new year with it. Thoughts?

Reviewer 4: 3rd of January SGTM. We can schedule it all in ahead of time.
I'm very happy to hear any official thoughts from CNCF on timing, too (if different).

> date: 2021-12-16

Reviewer 4: The date I'd use is 2021-12-20 - please update this (and adjust the path). That keeps the announcement separate from v1.23 post-release blogs.

Reviewer 4: See https://github.com/kubernetes/website/pull/30972#discussion_r770804291 for an update.

slug: kubernetes-is-moving-on-from-dockershim
---

**Authors:** Sergey Kanzhelev (Google), Jim Angel (Google), Davanum Srinivas (VMware), Shannon Kularathna (Google), Chris Short (AWS), Dawn Chen (Google)

> Kubernetes is removing dockershim in the upcoming v1.24 release. Please welcome new
> container runtimes with more exciting features, a smaller kubelet, and faster
> engineering velocity for new features. If you [use Docker Engine as a Container Runtime](/docs/tasks/administer-cluster/migrating-from-dockershim/find-out-runtime-you-use/)
> for your Kubernetes cluster, get ready to migrate from dockershim! To check if you're
> affected, refer to [Check whether dockershim deprecation affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/).

Reviewer 5: Per my previous comments, let's try and avoid any connotation that docker is "old", lacking in features, etc. Try: 

```suggestion
Kubernetes is removing dockershim in the upcoming v1.24 release. We're excited to reaffirm our community values by supporting open source container runtimes, enablin a smaller kubelet, and faster engineering velocity for teams using Kubernretes. If you [use Docker Engine as a Container Runtime](/docs/tasks/administer-cluster/migrating-from-dockershim/find-out-runtime-you-use/)
for your Kubernetes cluster, get ready to migrate to 1.24! To check if you're
affected, refer to [Check whether dockershim deprecation affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/).
```

Reviewer 4: Very much agreed. Typo in the suggestion BTW (“Kubernretes”).

## Why we’re moving away from dockershim

> Docker was the first container runtime used by Kubernetes. This is one of the
> reasons why Docker is so familiar to many Kubernetes users and enthusiasts.
> Later, we wanted to support other container runtimes. Today, Kubernetes uses the
> container runtime interface (CRI) to let system components (like the kubelet)
> talk to container runtimes. That original, baked in support for Docker turned
> into the component we call dockershim. It was built into Kubernetes code.
> Removing dockershim from Kubernetes inevitably affects the entire CNCF family of projects beyond
> Kubernetes. Dependencies on Docker and dockershim have crept into various tools
> and projects in the Kubernetes ecosystem.

Reviewer 5: ```suggestion
Docker was the first container runtime used by Kubernetes. This is one of the
reasons why Docker is so familiar to many Kubernetes users and enthusiasts. 
However, the project always wanted to support multiple runtimes, which is why we 
implemented the container runtime interface (CRI) to enable interoperability.
That said, Kubernetes still contained special code specifically for compatability 
with Docker, or the component we now know as dockershim. 
```
